README.md
## uxcore-progress

[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![Test Coverage][coveralls-image]][coveralls-url]
[![Dependency Status][dep-image]][dep-url]
[![devDependency Status][devdep-image]][devdep-url]
[![NPM downloads][downloads-image]][npm-url]

[![Sauce Test Status][sauce-image]][sauce-url]

[npm-image]: http://img.shields.io/npm/v/uxcore-progress.svg?style=flat-square
[npm-url]: http://npmjs.org/package/uxcore-progress
[travis-image]: https://img.shields.io/travis/uxcore/uxcore-progress.svg?style=flat-square
[travis-url]: https://travis-ci.org/uxcore/uxcore-progress
[coveralls-image]: https://img.shields.io/coveralls/uxcore/uxcore-progress.svg?style=flat-square
[coveralls-url]: https://coveralls.io/r/uxcore/uxcore-progress?branch=master
[dep-image]: http://img.shields.io/david/uxcore/uxcore-progress.svg?style=flat-square
[dep-url]: https://david-dm.org/uxcore/uxcore-progress
[devdep-image]: http://img.shields.io/david/dev/uxcore/uxcore-progress.svg?style=flat-square
[devdep-url]: https://david-dm.org/uxcore/uxcore-progress#info=devDependencies
[downloads-image]: https://img.shields.io/npm/dm/uxcore-progress.svg
[sauce-image]: https://saucelabs.com/browser-matrix/uxcore-progress.svg
[sauce-url]: https://saucelabs.com/u/uxcore


#### setup develop environment

```sh
$ git clone https://github.com/uxcore/uxcore-progress
$ cd uxcore-progress
$ npm install
$ npm start
```

## Usage
```js
import {Line, Circle} from 'uxcore-progress';
ReactDOM.render(<Line percent={30} />, target);
ReactDOM.render(<Circle status="active" />, target);
```

## demo
http://uxco.re/components/progress/

## API

## Props

### Line

| 配置项 | 类型 | 必填 | 默认值 | 功能/备注 |
|---|---|---|---|---|
|percent | 百分比 | number | 0 | |
|status | 状态，有两个值normal、exception、active三种状态 | string | normal | |
|strokeWidth | 进度条线的宽度，单位是px | number | 10 | |
|showInfo | 是否显示进度数值和状态图标 | bool | true | |

### Circle

| 配置项 | 类型 | 必填 | 默认值 | 功能/备注 |
|---|---|---|---|---|
|percent | 百分比 | number | 0 | |
|status | 状态，有两个值normal、exception、active三种状态 | string | normal | |
|strokeWidth | 进度条线的宽度，单位是px | number | 6 | |
|showInfo | 是否显示进度数值和状态图标 | bool | true | |
|size | 尺寸，相当于直径 | number | 156 | |

HISTORY.md
# History
## 1.3.3
- `NEW` add new prop infoType (icon|percent)
- `FIXED` fix state icon lost when success & exception in circle progress

## 1.3.1
- `FIXED` lost variable in less

## 1.3.0
- `CHANGED` support js style export

## 1.1.3
-`CHANGED` adapted React 15.x

## 1.1.2
- `FIXED` vertical align of the text with the line.

## 1.1.0
- `CHANGED` 1. apply new style 2. add circle progress

## 1.0.3
- `CHANGED` update scaffold

## 1.0.2
- `CHANGED` now can customize the icon

## 1.0.1
- `CHANGED` replace error & success icon with kuma-icon

## 1.0.0
- `NEW` first commit





tests/Circle.spec.js
import expect from 'expect.js';
import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Progress from '../src';

Enzyme.configure({ adapter: new Adapter() });

describe('<Progress.Circle />', () => {
  it('render default props', () => {
    const wrapper = mount(<Progress.Circle />);
    // percent
    expect(wrapper.props().percent).to.equal(0);
    // status
    expect(wrapper.props().status).to.equal('normal');
    expect(wrapper.find('.kuma-progress-circle-wrap').hasClass('status-normal')).to.equal(true);
    // strokeWidth
    expect(wrapper.props().strokeWidth).to.equal(6);
    // showInfo
    expect(wrapper.props().showInfo).to.equal(true);
    // size
    expect(wrapper.props().size).to.equal(156);
  });

  it('render 50 percent', () => {
    const wrapper = mount(<Progress.Circle percent={50} />);
    expect(wrapper.props().percent).to.equal(50);
  });

  it('render status is exception', () => {
    const wrapper = mount(<Progress.Circle status="exception" />);
    expect(wrapper.props().status).to.equal('exception');
    expect(wrapper.find('.kuma-progress-circle-wrap').hasClass('status-exception')).to.equal(true);
  });
  it('render status is active', () => {
    const wrapper = mount(<Progress.Circle status="active" />);
    expect(wrapper.props().status).to.equal('active');
    expect(wrapper.find('.kuma-progress-circle-wrap').hasClass('status-active')).to.equal(true);
  });

  it('render strokeWidth is 30', () => {
    const wrapper = mount(<Progress.Circle strokeWidth={30} />);
    expect(wrapper.props().strokeWidth).to.equal(30);
  });

  it('render showInfo is false', () => {
    const wrapper = mount(<Progress.Circle showInfo={false} />);
    expect(wrapper.props().showInfo).to.equal(false);
  });

  it('render 200 size', () => {
    const wrapper = mount(<Progress.Circle size={200} />);
    expect(wrapper.props().size).to.equal(200);
  });
});


tests/Progress.spec.js
import expect from 'expect.js';
import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Progress from '../src';

Enzyme.configure({ adapter: new Adapter() });

describe('<Progress.Line />', () => {
  it('render default props', () => {
    const wrapper = mount(<Progress.Line />);
    // all
    expect(wrapper.find('.kuma-progress-line-bg').html())
    .to.equal('<div class="kuma-progress-line-bg" style="width: 0%; height: 10px;"></div>');
    // percent
    expect(wrapper.props().percent).to.equal(0);
    // status
    expect(wrapper.props().status).to.equal('normal');
    expect(wrapper.find('.kuma-progress-line-wrap').hasClass('status-normal')).to.equal(true);
    // strokeWidth
    expect(wrapper.props().strokeWidth).to.equal(10);
    // showInfo
    expect(wrapper.props().showInfo).to.equal(true);
  });

  it('render 50 percent', () => {
    const wrapper = mount(<Progress.Line percent={50} />);
    expect(wrapper.props().percent).to.equal(50);
    expect(wrapper.find('.kuma-progress-line-bg').html())
      .to.equal('<div class="kuma-progress-line-bg" style="width: 50%; height: 10px;"></div>');
  });

  it('render status is exception', () => {
    const wrapper = mount(<Progress.Line status="exception" />);
    expect(wrapper.props().status).to.equal('exception');
    expect(wrapper.find('.kuma-progress-line-wrap').hasClass('status-exception')).to.equal(true);
  });
  it('render status is active', () => {
    const wrapper = mount(<Progress.Line status="active" />);
    expect(wrapper.props().status).to.equal('active');
    expect(wrapper.find('.kuma-progress-line-wrap').hasClass('status-active')).to.equal(true);
  });

  it('render strokeWidth is 30', () => {
    const wrapper = mount(<Progress.Line strokeWidth={30} />);
    expect(wrapper.props().strokeWidth).to.equal(30);
  });

  it('render showInfo is false', () => {
    const wrapper = mount(<Progress.Line showInfo={false} />);
    expect(wrapper.props().showInfo).to.equal(false);
  });
});


src/Progress.js
import React from 'react';
import PropTypes from 'prop-types';
import assign from 'object-assign';

class Line extends React.Component {

  render() {
    const props = assign({}, this.props);
    const prefixCls = props.prefixCls;

    if (parseInt(props.percent, 10) === 100) {
      props.status = 'success';
    }

    let progressInfo;
    let fullCls = '';

    if (props.showInfo === true) {
      const type = props.infoType
      if (props.status === 'exception') {
        progressInfo = (
          <span className={`${prefixCls}-line-text`}>
            {type === 'icon' ? <i className="icon-caution" /> : type === 'percent' ? `${props.percent}%` : 'Error'}
          </span>
        );
      } else if (props.status === 'success') {
        progressInfo = (
          <span className={`${prefixCls}-line-text`}>
            {type === 'icon' ? <i className="icon-success" /> : type === 'percent' ? `${props.percent}%` : 'Success'}
          </span>
        );
      } else {
        progressInfo = (
          <span className={`${prefixCls}-line-text`}>{props.percent}%</span>
        );
      }
    } else {
      fullCls = ` ${prefixCls}-line-wrap-full`;
    }

    const percentStyle = {
      width: `${props.percent}%`,
      height: props.strokeWidth,
    };

    const outerStyle = {
      paddingTop: (22 - props.strokeWidth) / 2,
    };

    return (
      <div className={`${prefixCls}-line-wrap fn-clear status-${props.status}${fullCls}`}>
        {progressInfo}
        <div className={`${prefixCls}-line-outer`} style={outerStyle}>
          <div className={`${prefixCls}-line-inner`}>
            <div className={`${prefixCls}-line-bg`} style={percentStyle}></div>
          </div>
        </div>
      </div>
    );
  }
}

Line.displayName = 'Progress-Line';
Line.propTypes = {
  prefixCls: PropTypes.string,
  status: PropTypes.oneOf(['normal', 'exception', 'active', 'success']),
  showInfo: PropTypes.bool,
  percent: PropTypes.number,
  strokeWidth: PropTypes.number,
  infoType: PropTypes.string
};
Line.defaultProps = {
  prefixCls: 'kuma-progress',
  percent: 0,
  strokeWidth: 10,
  status: 'normal', // exception active
  showInfo: true,
  infoType: 'icon'
};

export default Line;


demo/ProgressDemo.js
/**
 * Progress Component Demo for uxcore
 * @author vicent.bian
 *
 * Copyright 2014-2015, Uxcore Team, Alinw.
 * All rights reserved.
 */

import classnames from 'classnames';
import {Line, Circle} from '../src';

class Demo extends React.Component {

    constructor(props){
        super(props);
        this.state = {
            percent: 0
        };
    }

    increase() {
        let percent = this.state.percent + 10;
        if (percent > 100) {
            percent = 100;
        }
        this.setState({ percent });
    }
    decline() {
        let percent = this.state.percent - 10;
        if (percent < 0) {
            percent = 0;
        }
        this.setState({ percent });
    }

    render() {
        return (
            <div>
                <div>
                    <p>标准的进度条</p>
                    <Line percent={30} />
                    <Line percent={70} status="exception" />
                    <Line percent={100} />
                    <Line percent={50} showInfo={false} />
                    <Circle percent={30} />
                    <Circle percent={70} status="exception" infoType={'percent'} />
                    <Circle percent={100} status="success" />
                    <Circle percent={50} showInfo={false} />
                </div>
                <div style={{ width: 170 }}>
                    <p>适合放在较狭窄的区域内。</p>
                    <Line percent={30} strokeWidth={5} />
                    <Line percent={70} strokeWidth={5} status="exception" />
                    <Line percent={100} strokeWidth={5} />
                </div>
                <div>
                    <p>动态展示。</p>
                    <Line percent={this.state.percent} />
                    <Circle percent={this.state.percent} />
                    <button className="kuma-button kuma-button-secondary kuma-button-sm" onClick={this.decline.bind(this)}>
                        <i className="kuma-icon kuma-icon-stop"></i>
                    </button>
                    <button className="kuma-button-secondary kuma-button-sm" onClick={this.increase.bind(this)}>
                        <i className="kuma-icon kuma-icon-follow"></i>
                    </button>
                </div>
            </div>
        );
    }
};

module.exports = Demo;


src/Circle.js
import React from 'react';
import PropTypes from 'prop-types';
import assign from 'object-assign';

const defaultProps = {
  strokeWidth: 1,
  trailWidth: 1,
};

class Circle extends React.Component {

  render() {
    const props = assign({}, this.props);
    const prefixCls = props.prefixCls;
    const strokeWidth = props.strokeWidth;
    const radius = (50 - strokeWidth / 2);
    const pathString = `M 50,50 m 0,-${radius}
        a ${radius},${radius} 0 1 1 0,${2 * radius}
        a ${radius},${radius} 0 1 1 0,-${2 * radius}`;
    const len = Math.PI * 2 * radius;
    const pathStyle = {
      strokeDasharray: `${len}px ${len}px`,
      strokeDashoffset: `${((100 - props.percent) / 100 * len)}px`,
      transition: 'stroke-dashoffset 0.6s ease 0s, stroke 0.6s ease',
    };

    let progressInfo;
    let textStyle = {
      lineHeight: `${props.size}px`,

    };

    if (props.showInfo === true) {
      const type = props.infoType
      if (props.status === 'exception') {
        progressInfo = (
          <span style={textStyle} className={`${prefixCls}-circle-text`}>
            {type === 'icon' ? <i className="icon-caution" /> : type === 'percent' ? `${props.percent}%` : 'Error'}
          </span>
          );
      } else if (props.status === 'success') {
        progressInfo = (
          <span style={textStyle} className={`${prefixCls}-circle-text`}>
            {type === 'icon' ? <i className="icon-success" /> : type === 'percent' ? `${props.percent}%` : 'Success'}
          </span>
          );
      } else {
        progressInfo = (
          <span style={textStyle} className={`${prefixCls}-circle-text`}>{props.percent}%</span>
          );
      }
    }

    ['strokeWidth', 'trailWidth'].forEach((item) => {
      if (item === 'trailWidth' && !props.trailWidth && props.strokeWidth) {
        props.trailWidth = props.strokeWidth;
        return;
      }
      if (!props[item]) {
        props[item] = defaultProps[item];
      }
    });

    return (
      <div
        className={`kuma-progress-circle-wrap status-${props.status}`}
        style={{ width: props.size, height: props.size }}
      >

        <svg className="kuma-progress-circle" viewBox="0 0 100 100">
          <path
            className="kuma-progress-circle-trail"
            d={pathString}
            strokeWidth={props.trailWidth}
            fillOpacity="0"
          />
          <path
            className="kuma-progress-circle-path"
            d={pathString}
            strokeLinecap="round"
            strokeWidth={props.strokeWidth}
            fillOpacity="0"
            style={pathStyle}
          />
        </svg>
        {progressInfo}
      </div>
    );
  }
}

Circle.displayName = 'Progress-Circle';
Circle.propTypes = {
  prefixCls: PropTypes.string,
  status: PropTypes.oneOf(['normal', 'exception', 'success']),
  percent: PropTypes.number,
  strokeWidth: PropTypes.number,
  size: PropTypes.number,
  showInfo: PropTypes.bool,
  infoType: PropTypes.string
};
Circle.defaultProps = {
  prefixCls: 'kuma-progress',
  status: 'normal', // exception active
  percent: 0,
  strokeWidth: 6,
  size: 156,
  showInfo: true,
  infoType: 'icon'
};

export default Circle;
